Windows: zccache.exe exits STATUS_DLL_INIT_FAILED (-1073741502 / 0xC0000142) on CI runners

## Symptom

On Windows MSVC GitHub Actions runners, invoking `zccache` (observed via `zccache stop` in the setup-soldr post step) exits with code `-1073741502` = `0xC0000142` = **`STATUS_DLL_INIT_FAILED`**. This means the binary cannot complete DLL initialization during process startup — the Windows loader rejects the image before any of our code runs.

## Observed failures

zackees/soldr PR #308 (rebased + merged 2026-05-15) had both Windows legs of the cross-repo `build` workflow fail in this way despite the PR only touching `.github/workflows/release-auto.yml`:

- x86_64-pc-windows-msvc: https://github.com/zackees/soldr/actions/runs/25907556419/job/76143968484
- aarch64-pc-windows-msvc: https://github.com/zackees/soldr/actions/runs/25907556419/job/76143968433

Failing step is `setup-soldr@dc7390d` post-action `zccache stop`:

```
& $zccache stop
##[error]Process completed with exit code -1073741502.
Cleaning up orphan processes
Terminate orphan process: pid (9168) (zccache-daemon.697979224)
```

## Why this is a blocker

The Windows MSVC legs of every CI run that uses `setup-soldr@dc7390d` (or newer) are broken by this. The `zccache` binary is *installed* (the orphan-process cleanup confirms `zccache-daemon` was spawned), so the *daemon* started fine — but a subsequent `zccache <subcommand>` invocation can't get past loader initialization. That asymmetry is a clue.

## Hypotheses (rank-ordered)

1. **Transitive native dep added a delay-load or import that the runner image can't satisfy.** E.g. recent OpenSSL/`reqwest` bump, `windows-sys`/`windows` feature surface, or a Rust target that started emitting a delay-load directive. The daemon may not exercise the same code path until the symbol is touched.
2. **Build-time vs. run-time MSVC runtime mismatch** — binary built against a newer vcruntime/UCRT than the runner has staged. Unlikely on `windows-2022`/`2025` images but possible if we link against vcpkg-installed libs that pulled in a newer runtime stub.
3. **A new manifest entry / SxS dependency** appended to the exe.
4. **Antivirus/Defender quarantining the binary mid-load** — would generally produce a different error, but possible.

## What to reproduce

1. Spin up a `windows-2022` (or windows-2025) GitHub Actions self-hosted runner equivalent, OR clone the runner image locally via `microsoft/vhd-runner-images`.
2. Install zccache via the same path setup-soldr uses (download the published Windows wheel/binary that setup-soldr@dc7390d resolves).
3. Confirm `zccache --version` and `zccache stop` exit non-zero with `-1073741502`.
4. Run the binary under Process Monitor or `gflags +sls` (loader snaps) to get the exact missing/failing DLL.
5. `dumpbin /dependents zccache.exe` to enumerate import-table deps. Compare against the published `windows-2022` image manifest.

## Resolution path

- Identify the offending DLL via loader-snaps + `dumpbin`. Then either:
  - Static-link the library (preferred for a self-contained binary like this).
  - Pin the upstream dep version that doesn't introduce the load-time symbol.
  - Bundle the DLL alongside the exe in the release artifact.
- **Self-diagnose:** wrap startup in a Win32-aware error path. Before `main` runs, the loader has already failed — so we can't `eprintln!` from inside zccache. Instead, the **release pipeline** should run `zccache --version` on the same Windows image during a smoke step; if it can't even print its version, fail the release and avoid shipping a broken binary.
- Add a Windows smoke gate to the release workflow (`run: zccache --version` on a clean `windows-2022` image) so this regression cannot ship to PyPI.

## Acceptance criteria

- `zccache --version` and `zccache stop` exit `0` on clean `windows-2022` and `windows-2025` GitHub Actions runners.
- A new release-gate CI job runs the smoke check on every Windows build target and blocks publishing if it fails.
- Root cause is documented in the PR description (which DLL, which dep introduced it, how it's now resolved).

🤖 Filed via Claude Code

--- a/.github/actions/build-target/action.yml
+++ b/.github/actions/build-target/action.yml
@@ -98,14 +98,54 @@ runs:
         TARGET_UPPER=$(echo "${{ inputs.target }}" | tr '[:lower:]-' '[:upper:]_')
         echo "CARGO_TARGET_${TARGET_UPPER}_LINKER=${{ inputs.linker }}" >> "$GITHUB_ENV"
 
+    # Static-link the MSVC C runtime (vcruntime + UCRT) for Windows release
+    # binaries. Without `+crt-static` the published `zccache.exe` declares
+    # imports against `VCRUNTIME140.dll` and the `api-ms-win-crt-*` UCRT
+    # API-set forwarders; on runner images where any of those DLLs is
+    # missing or version-skewed, the Windows loader rejects the image
+    # before `main` with `STATUS_DLL_INIT_FAILED` (exit `-1073741502` /
+    # `0xC0000142`). Self-contained CRT linkage is the standard approach
+    # for distributable Rust binaries on Windows (rustc/cargo/rustup all
+    # ship this way). See zccache#269.
+    #
+    # We DO NOT apply +crt-static to the Python extension (.pyd) builds
+    # below — those load into the host Python process and must share its
+    # C runtime. Static CRT linkage is appended to RUSTFLAGS only for the
+    # standalone binary build step.
     - name: Build release binaries
       shell: bash
       env:
         PYO3_PYTHON: python
       run: |
+        if [[ "${{ inputs.target }}" == *pc-windows-msvc ]]; then
+          export RUSTFLAGS="${RUSTFLAGS:-} -C target-feature=+crt-static"
+        fi
         soldr cargo build --release --target ${{ inputs.target }} -p zccache-cli
         soldr cargo build --release --target ${{ inputs.target }} -p zccache-daemon
         soldr cargo build --release --target ${{ inputs.target }} -p zccache-fingerprint
+
+    # Windows DLL load smoke gate (see zccache#269). If the just-built
+    # `zccache.exe` cannot complete loader init on this host, `--version`
+    # exits non-zero before any Rust code runs and the release stops here
+    # rather than shipping a binary that will fail in CI consumers. Other
+    # platforms run the same check for symmetry but the Windows leg is
+    # the one that actually catches the regression. ARM64 Windows builds
+    # are cross-compiled on x64 runners and so cannot exec the binary on
+    # this leg — skip the smoke for that target only.
+    - name: Smoke test built binary
+      shell: bash
+      run: |
+        case "${{ inputs.target }}" in
+          aarch64-pc-windows-msvc|aarch64-unknown-linux-musl|aarch64-unknown-linux-gnu)
+            # Cross-compiled; skip exec on host. TODO: when matrix gains
+            # an arm64 runner, gate aarch64-pc-windows-msvc through it.
+            echo "Skipping smoke test for cross-compiled target ${{ inputs.target }}"
+            exit 0
+            ;;
+        esac
+        BIN="target/${{ inputs.target }}/release/zccache${{ inputs.binary_ext }}"
+        echo "Smoke test: $BIN --version"
+        "$BIN" --version
 
     - name: Build Python extension artifacts
       if: inputs.include_python == 'true'

